### PR TITLE
[FLINK-2093][gelly] Added difference method

### PR DESCRIPTION
**Pull request by shghatge**
Tasks given on 5th June:
Add a difference function to the Graph.java
Modify the docs 'gelly-guide.md'
Add the test case for difference() method to GraphMutationsITCase.java

### CODE REVIEW

File: docs/libs/gelly_guide.md [hunk 1]
```diff
@@ -236,6 +236,8 @@ Graph<Long, Double, Double> networkWithWeights = network.joinWithEdgesOnSource(v
 
 * <strong>Union</strong>: Gelly's `union()` method performs a union on the vertex and edges sets of the input graphs. Duplicate vertices are removed from the resulting `Graph`, while if duplicate edges exists, these will be maintained.
 
+* <strong>Difference</strong>: Gelly's `difference()` method performs a difference on the vertex and edges sets of the input graphs. Common vertices are removed from the resulting `Graph`, along with the edges which which have these vertices as source/target.
```
Comment (andralungu) [inline docs/libs/gelly_guide.md]: you have written which twice, "along with the edges which which" :)

File: flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/Graph.java [hunk 2]
```diff
@@ -1233,6 +1233,34 @@ public void coGroup(Iterable<Edge<K, EV>> edge, Iterable<Edge<K, EV>> edgeToBeRe
 		return new Graph<K, VV, EV>(unionedVertices, unionedEdges, this.context);
 	}
 
+    /**
+     * Performs Difference on the vertices and edges sets of the inputgraphs
```
Comment (andralungu) [inline flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/Graph.java]: "on the vertex and edge sets of the input graphs"

File: flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/Graph.java [hunk 3]
```diff
@@ -1233,6 +1233,34 @@ public void coGroup(Iterable<Edge<K, EV>> edge, Iterable<Edge<K, EV>> edgeToBeRe
...
+     * removes both vertices and edges with the vertex as a source/target
+     * @param graph the graph to perform differennce with
+     * @return a new graph
```
Comment (andralungu) [inline flink-staging/flink-gelly/src/main/java/org/apache/flink/graph/Graph.java]: a new graph where the common vertices and edges have been removed

File: flink-staging/flink-gelly/src/test/java/org/apache/flink/graph/test/operations/GraphOperationsITCase.java [hunk 4]
```diff
@@ -266,6 +266,47 @@ public void testUnion() throws Exception {
...
+		/*
+		 * Test difference()
+		 */
+        final ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
+
+        Graph<Long, Long, Long> graph = Graph.fromDataSet(TestGraphUtils.getLongLongVertexData(env),
+                TestGraphUtils.getLongLongEdgeData(env), env);
+
+        List<Vertex<Long, Long>> vertices = new ArrayList<Vertex<Long, Long>>();
+        List<Edge<Long, Long>> edges = new ArrayList<Edge<Long, Long>>();
```
Comment (andralungu) [inline flink-staging/flink-gelly/src/test/java/org/apache/flink/graph/test/operations/GraphOperationsITCase.java]: I would put these in TestGraphUtils, one remove is fine, but three can make the code  a bit difficult to read :)

File: flink-staging/flink-gelly/src/test/java/org/apache/flink/graph/test/operations/GraphOperationsITCase.java [hunk 5]
```diff
@@ -266,6 +266,47 @@ public void testUnion() throws Exception {
...
+        vertices.remove(4);
+
+        vertices.add(new Vertex<Long,Long>(6L,6L));
+
+        edges.remove(0);
+        edges.remove(2);
+        edges.remove(3);
+        edges.remove(4);
+        edges.remove(5);
+        edges.remove(6);
```
Comment (andralungu) [inline flink-staging/flink-gelly/src/test/java/org/apache/flink/graph/test/operations/GraphOperationsITCase.java]: same for the edges

File: flink-staging/flink-gelly/src/test/java/org/apache/flink/graph/test/operations/GraphOperationsITCase.java [hunk 6]
```diff
@@ -266,6 +266,47 @@ public void testUnion() throws Exception {
...
+
+        edges.add(new Edge<Long, Long>(6L,1L,61L));
+        edges.add(new Edge<Long, Long>(6L,3L,63L));
+
+        graph = graph.difference(Graph.fromCollection(vertices, edges, env));
+
+        graph.getEdges().writeAsCsv(resultPath);
+        graph.getVertices().writeAsCsv(resultPath);
```
Comment (andralungu) [inline flink-staging/flink-gelly/src/test/java/org/apache/flink/graph/test/operations/GraphOperationsITCase.java]: The graph.getVertices() should actually be in a different test; that way you could change the expected result and see that the vertices you get are actually the ones you expected.

**Comment by andralungu** [comment]
Hi @shghatge ,

This very nice for a first PR and I am happy to see that you followed my guidelines :)
I left a set of comments in-line.

Apart from those:
- the difference method can be simplified. You don't need to filterOnEdges. Have a closer look at removeVertices. Imagine what happens if you remove a vertex, the edge will also have to be removed. You cannot leave an edge with just the source or the target vertex trailing.
- I think you forgot to add the corner case test for an input graph which does not have common vertices with the first one. I know you wrote it :) 
-  Finally, if you have a look at the Travis build here, it failed because you are indenting with spaces instead of tabs. You should play a bit with your IntelliJ settings. No worries! This is a rookie mistake, we all did it at first. To check everything is okay, just do a cd flink-staging/flink-gelly and then mvn verify. After it says build success, we're good to go. 
Rebase and update the PR.

If you have questions, I'll be more than happy to answer them!  Nice job!

**Comment by shghatge** [comment]
Then it was just removing vertices! Talk about swatting a Fly with a Sledgehammer! I will do all the changes you suggested.

**Comment by fhueske** [comment]
Hi @shghatge, you don't need to close a PR in order to update it.
You can simply update (push or push --force into) the branch from which you created the PR and Github will automatically update the PR. This helps to have all comments about your implementation in one place.

**Comment by shghatge** [comment]
Hi @fhueske I am sorry for closing the pull request. Now onwards I will follow the approach you suggested.

**Comment by fhueske** [comment]
@shghatge no worries. It's not a big deal :-)

**Pull request closed without merging**